Add directive support and mandatory @skip and @include

anmonteiro opened the pull request:
fixes #23

rrosenzweig commented on graphql/src/graphql_schema.ml:
```diff
@@ -356,41 +360,33 @@ module Make (Io : IO) (Stream: Stream with type 'a io = 'a Io.t) = struct
+      typ        : ('ctx, 'out) typ;
+      args       : ('out, 'args) Arg.arg_list;
+      resolve    : 'ctx -> 'args;
+    } -> ('ctx, 'out) directive
```
To simplify, I would propose:

```ocaml
  type directive =
    Directive : {
      name       : string;
      doc        : string option;
      locations  : directive_location list;
      args       : (bool, 'args) Arg.arg_list;
      resolve    : 'args;
    } -> directive
```

rrosenzweig commented on graphql/src/graphql_schema.ml:
```diff
@@ -437,14 +433,6 @@ module Make (Io : IO) (Stream: Stream with type 'a io = 'a Io.t) = struct
-      fields
-    in
-    { name; doc; abstracts = ref []; fields = lazy fields }
-
```
What are your thoughts on moving these lines (and the ones above) further down?

anmonteiro replied:
Where exactly? I've moved them just before the introspection module which feels ok to me.

rrosenzweig replied:
Sorry, I was being unclear: why move them at all in this PR? 😄

anmonteiro replied:
Oh we actually don't need to. I was envisioning passing some directives argument to `schema`, which would mean their definition would have to come before.

rrosenzweig commented on graphql/src/graphql_schema.ml:
```diff
@@ -1048,26 +1179,65 @@ end
+        let includ_args = includ.arguments in
+        Arg.eval_arglist ctx.variables skip_directive.args skip_args skip_resolver >>= fun skip_res ->
+        Arg.eval_arglist ctx.variables include_directive.args includ_args include_resolver >>| fun include_res ->
+          not skip_res && include_res
```
I would propose a recursive definition along these lines:

```ocaml
  let rec should_include ctx (directives : Graphql_parser.directive list) =
    match directives with
    | [] -> Ok true
    | { name = "skip"; arguments }::rest ->
        (* ... *)
    | { name = "include"; arguments }::rest ->
        (* ... *)
    | { name; _ }::_ ->
        let err = Format.sprintf "Unknown directive: %s" name in
        Error err
```

rrosenzweig commented on graphql/src/graphql_schema.ml:
```diff
@@ -1048,26 +1179,65 @@ end
     List.map (function
     | Graphql_parser.Field field ->
-        [field]
+        should_include ctx field.directives >>| fun should_include ->
```
I think it's unfortunate with the name collision.

anmonteiro commented:
Addressed your comments and rebased on master.

rrosenzweig commented on graphql/src/graphql_schema.ml:
```diff
@@ -7,6 +7,10 @@ module List = struct
+  let rec find_opt p = function
+    | [] -> None
+    | x :: l -> if p x then Some x else find_opt p l
+
```
This seems unused now.

rrosenzweig commented on graphql/src/graphql_schema.ml:
```diff
@@ -356,41 +360,33 @@ module Make (Io : IO) (Stream: Stream with type 'a io = 'a Io.t) = struct
+      name       : string;
+      doc        : string option;
+      locations  : directive_location list;
+      typ        : ('ctx, 'out) typ;
```
This field is no longer required, I think.

rrosenzweig commented on graphql/src/graphql_schema.ml:
```diff
@@ -1068,26 +1199,54 @@ end
+          let Directive directive = include_directive in
+          let resolver = directive.resolve in
+          Arg.eval_arglist ctx.variables directive.args arguments resolver
+          |> Rresult.R.map (fun includ -> includ && should_include_rest)
```
Instead of recursing first and considering the current element later, how considering the current element first? This allows you to short-circuit in case the current element indicates the field should not be included.

rrosenzweig commented on graphql/src/graphql_schema.ml:
```diff
@@ -1068,26 +1199,54 @@ end
     List.map (function
     | Graphql_parser.Field field ->
-        [field]
+        should_include ctx field.directives >>| fun should_include_field ->
```
I would suggest giving the variable the short name and the function the longer name. Maybe the function `should_include_field` and the variable `include_field`?

rrosenzweig commented on graphql/src/graphql_schema.ml:
```diff
@@ -476,6 +464,67 @@ module Make (Io : IO) (Stream: Stream with type 'a io = 'a Io.t) = struct
+    args = Arg.[
+      arg "if" ~doc:"Skipped when true." ~typ:(non_null bool)
+    ];
+    resolve = fun if_ -> if_
```
Maybe return `not if_` here to make the directives homogenous in the semantics of their return value (i.e. whether the field should be included or not)?

anmonteiro replied:
Interesting thought – I think that conflates the directive definition with its usage in `should_include`, though. 

I'm imagining that we could want the raw value of `"skip"` in some future usage (i.e. if the spec evolves).

andreas replied:
What if the return type was ```[`Skip | `Include]```?

anmonteiro commented:
Reworked the PR to include your latest comments.

anmonteiro commented:
I had forgotten about this PR until I watched @sgrove's talk at Øredev 2018 where he mentioned directives were still WIP 😅 

I just rebased this on master and fixed the tests. @andreas could you take another look at this please?

anmonteiro commented:
@andreas addressed the latest review comments and rebased on master.

andreas commented:
Thanks a lot! 🏆
